### PR TITLE
Expose memory API & create host function DSL

### PR DESCRIPTION
**Pull request by alexdovzhanyn**
Solves #25 

## Host function DSL usage
An extensive example of the new host function DSL usage can be found [here](https://github.com/alexdovzhanyn/WebAssembly_TicTacToe), but the basic usage is as follows:

When importing a host function into a WebAssembly, the text representation expects an expression in the format (`?` denoting optional fields):
```webassembly
(import "module_name" "function_name" (func ?(param [type]) ?(result [type])))
```

A module name is required, as well as a function name. This maps pretty well to Elixir, since you can't have functions without them being contained in a module. The DSL takes advantage of this by allowing users to namespace Elixir modules as WebAssembly modules. For example:

```elixir
defmodule Math do
  use WaspVM.HostFunction # Allows usage of the defhost macro

  # Defines a host function
  defhost :add, [a, b] do
    a + b
  end
end
```

When the above is passed as an import into the VM, it'll create a host function that's accessible within a WebAssembly module with the following expression:

```webassembly
(import "Math" "add" (func (param i32 i32) (result i32))
```

As you can see, the elixir module name maps directly to the WebAssembly module name. The `defhost` macro specifies exactly which functions should be exported to WebAssembly -- this makes it possible to mix and match normal elixir functions along with host functions:

```elixir
defmodule Greeter do
  use WaspVM.HostFunction

  defhost :get_name_and_greet do
    name = get_name()
    greet(name)
  end

  def get_name do
    IO.gets "What is your name?"
  end

  def greet(name) do
    IO.puts "Hello there, #{name}"
  end
end
```

In the above example, only a host function called "get_name_and_greet" is defined in the VM; `get_name` and `greet` are only namespaced within the elixir module. This means you could define a host function called "get_name" and it would be a completely different function from `Greeter.get_name/0`

The syntax for the `defhost` macro is as follows:

* The function name can be either an atom or a string. If it's an atom, it'll get converted to a string internally.
* If the function accepts parameters, those can be accessed via a list pattern match directly following the function name. **There must be a comma separating the function name and the list**.
* If the function doesn't accept parameters, the function name should be followed by a `do` block.

## Using the new API

In this PR, the only functionality in the API is access to memory, but more features may be added later.

To access the API from within a host function, there is a variable within the scope of every host function called `ctx`. This variable is used as an input into the API functions defined in the `WaspVM.HostFunction.API` module. For example, to read memory in a host function:

```elixir
defhost :read_mem do
  WaspVM.HostFunction.API.get_memory(ctx, "main_mem", 0, 32)
end
```

The above will define a host function that reads 32 bytes from an **exported** (exported from within the Wasm module) memory called "main_mem", starting at memory address 0.

### CODE REVIEW

File: lib/decoding/import_section_parser.ex [hunk 1]
```diff
@@ -10,7 +10,7 @@ defmodule WaspVM.Decoder.ImportSectionParser do
 
     entries = if count > 0, do: parse_entries(entries), else: []
 
-    {:imports, entries}
+    {:imports, Enum.reverse(entries)}
```
Comment (fantypants) [inline lib/decoding/import_section_parser.ex]: Just a quick note, but we should check to see how many of these types of functions we're using, if were duplicating the function calls, we may be able to reduce execution time through refactoring the data set

**Comment by OvermindDL1** [comment]
Why not use the `defhost add(a, b) do ... end` form instead of `defhost :add, [a, b] do ... end`, which is significantly more following Elixir's style?

**Comment by alexdovzhanyn** [comment]
@OvermindDL1 We use the arguments passed to defhost to define a function within the module using those same arguments. As far as I know, there isn't a way to create a macro that allows any arity arguments to be passed in and then use those to be passed into a def call. 

What we're doing is:

```elixir
defmacro defhost(fname, args, do: block) do
  quote generated: true do
    def hostfunc(unquote(fname), unquote(args), var!(ctx)), do: unquote(block)

    ...
  end
end
```

These later get pattern matched against when the function is compiled, and we need to know a common function name to do this (hence the definition of every function as a variant of `hostfunc`.

If you know of a way to convert this to use your above syntax, please share!

**Comment by OvermindDL1** [comment]
> If you know of a way to convert this to use your above syntax, please share!

Precisely what I put.  :-)

Using your example but fixing it up:
```elixir
defmacro defhost(head, do: block) do
  {fname, args} = Macro.decompose_call(head)
  quote generated: true do
    def hostfunc(unquote(fname), unquote(args), var!(ctx)), do: unquote(block)
      ...
    end
  end
end
```
And that will use the `defhost blah(a, b, c) do ... end` form.  It's basically the way `def`/`defp`/etc... are defined as well as how other libraries do it, super common pattern.  :-)

**Comment by alexdovzhanyn** [comment]
@OvermindDL1 That's perfect, looks _much_ cleaner, thank you! This was impossible to query for so I didn't even know it was possible. I'll add this in the next commit.

**Comment by OvermindDL1** [comment]
Hehe, awesome!  If you want to support guards as well it's a touch bit more work but still easy as well.  :-)

**Comment by alexdovzhanyn** [comment]
@OvermindDL1  I'll look into adding guard support in a future version, but I feel that it's a tad less critical at this point. You made this suggestion right in time, I've added the changes and released it with v0.8.1 :)

**Comment by OvermindDL1** [comment]
Awesome, well if you want guards then adjusting the top example it would look similar to this:
```elixir
defmacro defhost({:when, _when_ctx, [head, guard]}, do: block) do
  {fname, args} = Macro.decompose_call(head)
  gen_host(fname, args, guard, do_body)
end
defmacro defhost(head, do: block) do
  {fname, args} = Macro.decompose_call(head)
  gen_host(fname, args, true, do_body)
end

defp gen_host(fname, args, guard, do_body) do
  quote generated: true do
    def hostfunc(unquote(fname), unquote(args), var!(ctx)) when unquote(guard), do: unquote(block)
      ...
    end
  end
end
```
Or invert it and pull out the when/head decomposition into its own function, whichever.  :-)

**Pull request merged**